src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+        let this = self.project();
+        match this.future.poll(cx) {
+            Poll::Ready(v) => {
+                this.state.take().unwrap().unsubscribe();
+                Poll::Ready(v)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// A future that blocks until a graceful shutdown is complete.
+pub struct GracefulWaiter {
+    state: GracefulWaiterState,
+}
+
+impl GracefulWaiter {
+    fn new(state: Arc<GracefulState>) -> Self {
+        Self {
+            state: GracefulWaiterState { state, key: None },
+        }
+    }
+}
+
+impl Future for GracefulWaiter {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let state = &mut self.state;
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            return Poll::Ready(());
+        }
+
+        let mut waker_list = state.state.waker_list.lock().unwrap();
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            // check again in case of race condition
+            return Poll::Ready(());
+        }
+
+        let waker = Some(cx.waker().clone());
+        state.key = Some(match state.key.take() {
+            Some(key) => {
+                *waker_list.get_mut(key).unwrap() = waker;
+                key
+            }
+            None => waker_list.insert(waker),
+        });
+
+        Poll::Pending
+    }
+}
+
+struct GracefulWaiterState {
+    state: Arc<GracefulState>,
+    key: Option<usize>,
+}
+
+impl Drop for GracefulWaiterState {
+    /// When the waiter is dropped, we need to remove its waker from the waker list.
+    /// As to ensure the graceful waiter is cancel safe.
+    fn drop(&mut self) {
+        if let Some(key) = self.key.take() {
+            let mut wakers = self.state.waker_list.lock().unwrap();
+            wakers.remove(key);
+        }
+    }
+}
+
+pin_project! {
+    struct GracefulConnectionFuture<C, F> {
+        #[pin]
+        conn: C,
+        #[pin]
+        cancel: F,
+        cancelled: bool,
+    }
+}
+
+impl<C, F> GracefulConnectionFuture<C, F> {
+    fn new(conn: C, cancel: F) -> Self {
+        Self {
+            conn,
+            cancel,
+            cancelled: false,
+        }
+    }
+}
+
+impl<C, F> Debug for GracefulConnectionFuture<C, F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GracefulConnectionFuture").finish()
+    }
+}
+
+impl<C, F> Future for GracefulConnectionFuture<C, F>
+where
+    C: GracefulConnection,
+    F: Future,
+{
+    type Output = C::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        if !*this.cancelled {
+            if let Poll::Ready(_) = this.cancel.poll(cx) {
```

Comment:
`is_ready` can be used, and then this nested `if` statement can be flattened.

Reply:
ok will check it out